src/crimson/net/ProtocolV2.cc

```diff
@@ -1982,9 +1976,7 @@ seastar::future<> ProtocolV2::read_message(utime_t throttle_stamp)
 
     // TODO: change MessageRef with seastar::shared_ptr
     auto msg_ref = MessageRef{message, false};
-    gated_dispatch("ms_dispatch", [this, msg = std::move(msg_ref)] {
-      return dispatcher.ms_dispatch(&conn, std::move(msg));
-    });
+    (void)dispatcher.ms_dispatch(&conn, std::move(msg_ref));
```

Comment:
i think it's fine though. we do the same everywhere else. and i cannot see why we should change this.

Reply:
I mean there's a missing space there.

Reply:
yes, i don't think we need to have space after `(void)`.

Reply:
A space in C-style cast? I would vote against...

Reply:
`std::ignore = ...` instead? Assignment looks better than the braces IMHO.

Reply:
The gated_dispatch bit here is what prevented the next message from being dispatched before this one completes, I think.  I'm not following how removing it doesn't permit message reordering.

Reply:
@athanatos I don't quite follow, gated_dispatch also return void. Are you suggesting this could be a problem that have existed for a long time? And I think if op A arrives at OSDs before the ack of op B, these two ops can be considered concurrent, does the order of this kind of ops matter?

Reply:
@athanatos Oh, it matters when ops are committed on both primary and replica osds, right?

Reply:
@cyx1231st @athanatos @rzarzynski @tchaikov @ronen-fr So, basically, we have to

1. keep those "ms_XXX" method return seastar::future<>;
2. resolve the op ordering problem.

Is this right?

Reply:
I misunderstood pending_dispatch in gated_dispatch, I think.  However, this is a real problem.  If the primary sends two ops in order to the replica, the metadata component of the two transactions *must* be applied in order, or none of the data safety guarantees work.  You likely noticed one of these assumptions in working on log based recovery -- we assume that if two osds find the same last_update log version on disk they must both have applied the same previous updates.

@cyx1231st I thought we ensured that the dispatchers were invoked synchronously to avoid this?

Reply:
@athanatos I think the order of these ops are guaranteed in ClientRequest::start() by the blocking_futures, is this right?

Reply:
Yeah, those are my primary concerns.  Keep in mind, the dispatcher can itself immediately fork off an async process as appropriate to free up the messenger to read another message.  It's just critical that the messenger give the dispatcher the opportunity to at least look at the messages in order.

Reply:
That is, start_operation will invoke RepRequest::start() which invokes with_blocking_future on await_map.  The op is placed into the queue before start() itself returns a future.

Conceivably, calling the dispatcher asyncronously might work, but it relies on no intervening calls returning a future that actually blocks before start() is called.  That seems brittle.  It also prevents the dispatcher from actually stopping reading the next message -- that might actually be desirable sometimes.

Reply:
> Only if start() is called in order. That's the problem, with this implementation the second message might get to start() first.

I think the ms_dispatch would synchrously invoke ClientRequest::start(), so as long as ms_dispatch here are invoked in order, problem solved, right?

Reply:
The other option would be to insist that all dispatchers simply return void.  In that case, the chained ms_dispatch would just use a synchronous for loop rather than seatar::do_for_each.  Done that way, the implicit contract would be that Dispatcher handlers may never block (and therefore may never perform IO).  Of course, they could still initiate an aync operation as most do now.

I kind of prefer the idea of allowing the dispatcher methods block the reader though.  It gives us a way of implementing back pressure.

Reply:
Doing it that way would probably simplify shutdown, too.

Reply:
@xxhdx1985126 That might *happen* to be true at the moment, but it relies on your seastar::do_for_each call not returning a future prior to invoking that dispatcher.  I think that's true at the moment, but it's not a property of the do_for_each contract, and a future modification might add something that actually does block (like waiting on some kind of dispatch throttle, for instance).

As this is critically important to correctness (and incredibly hard to debug when it breaks), I'd like us to be explicit by either chaining the continuations correctly, or requiring the handlers to be fully synchronous.

Reply:
> @cyx1231st I thought we ensured that the dispatchers were invoked synchronously to avoid this?

If you are referring to https://github.com/ceph/ceph/pull/28321, I think we haven't decided to merge it. So currently a later message will be delivered before the previous `ms_dispatch()` finished, because their futures are not chained.

Reply:
@athanatos @cyx1231st You mean we make Dispatcher::ms_XXXX return void and ChainedRequests::ms_XXXX do simple "for" rather than "do_for_each", right? If so, I think I got your point, but right now, I don't quite understand how ChainedDispatchers::ms_dispatch are issued in the order of the arrival of messages in the first place, it seems the following code involve the possible out-of-order invocation of ChainedDispatcher::ms_dispatch:
```
seastar::future<> ProtocolV2::read_message(utime_t throttle_stamp)
{
  return read_frame_payload()
  .then([this, throttle_stamp] {
    utime_t recv_stamp{seastar::lowres_system_clock::now()};

```

Reply:
Got your point.

Reply:
It seems you are right, the current implementation is very brittle.  It only works because nothing returns a blocking future prior to calling the osd dispatcher.  Adding a dispatcher that actually does block ahead of the osd I think would break this property.  From https://github.com/ceph/ceph/pull/28321 , it appears that we decided to put off fixing it.  @cyx1231st said at the time:

Update: this is what we want in the future, but no now.. depend on the progress and requirement of crimson OSD developments.

Since then, we've implemented protocols that certainly do depend on that message ordering property.  IMHO, this patch makes the brittleness more difficult to fix by preventing ChainedDispatcher from propagating the future.  At a minimum, we should preserve the existing behaviour so that we can we more easily fix it in the future.

Reply:
> The other option would be to insist that all dispatchers simply return void. In that case, the chained ms_dispatch would just use a synchronous for loop rather than seastar::do_for_each. Done that way, the implicit contract would be that Dispatcher handlers may never block (and therefore may never perform IO). Of course, they could still initiate an aync operation as most do now.

If I remember correctly from #28321, we decided to go with this option without back pressure, but still keep `ms_dispatch()` to return future. So `seastar::do_for_each` is also not correct for `ms_dipatch()` under that scope.

Reply:
You are right, the current implementation is very brittle.  It only works because nothing returns a blocking future prior to calling the osd dispatcher.  Adding a dispatcher that actually does block ahead of the osd I think would break this property as the do_for_each would block before executing the osd handler.

From your comment on https://github.com/ceph/ceph/pull/28321, it looks like we were just putting off fixing it.  Since then, we've implemented protocols that certainly do depend on that message ordering property.  IMHO, this patch makes the brittleness more difficult to fix by preventing ChainedDispatcher from propagating the future.  I think at least we should preserve the property that ChainedDispatcher propagates the futures.

I also think we should prioritize clarifying this interface.  If ms_dispatch (and the other handlers) can return a future that blocks, we need to respect that by blocking the read sequence.  If we expect them never to block (as we essentially do with this implementation), we should update the interface to reflect that by not returning a future.

Reply:
@athanatos @cyx1231st I think, to summarize the above discussion, we can make ChainedRequests::ms_dispatcher do "for" rather than "do_for_each" and return a future<> that are resolved when all the futures returned by Dispatcher::ms_dispatch are resolved, while keep Dispatcher::ms_XXX() methods returning future<>, so different events on the same connection would be handled in order. Is this right?

Reply:
No.  That would leave us where we are now where we cannot guarantee that any particular handler gets called before the next message gets read unless the interface is used in a very narrow fashion.  I'm suggesting either:

1) Dispatcher::ms_* and ChainedDispatchers::ms_* both return void.

In this case, ChainedDispatchers::ms_* can simply iterate over the dispatchers and call the corresponding methods.  As the contained Dispatcher handlers do not return futures (and cannot block), ChainedDispatchers handlers need not either.

2) Dispatcher::ms_* and ChainedDispatchers::ms_* both return seastar::future<>

Here, because Dispatcher methods may block, ChainedDispatcher methods must also return seastar::future<>.  For this PR, there are then two options:
a) Ignore the returned future in ProtocolV2 for now and fix it later.  This leaves us in the same state as now where adding a blocking dispatcher or otherwise failing to run all the way through to the osd handler would break the message ordering guarantees.
b) Actually block the ProtocolV2 reader until the ChainedDispatcher handler resolves.

I'm sorry for being insistent about this.  I think that ordering guarantees in distributed systems are absolutely critical to being able to reason about the correctness properties of the system, and so it's worth being clear about what they are so that we don't inadvertently break them later.

Reply:
@athanatos I'm still a little confused, if we make ms_dispatch return void and other ms_* return future<>, and make Protocol those future<>, then, as long as ms_dispatch is not blocked before reaches other synchonization structures, like in OSD::ms_dispatch,  we don't have a problem. Is this right?

Reply:
@athanatos The future returned by `ms_dispatch()` is meaningful to block connection reads, and inherently there is a TCP window support through the other end. But what about other types of events?

For example `ms_dispatch_reset` returns a future back to messenger, is it meaningful? My understanding is that reset events work like interrupts that signals application in real time that the connection is lost. So seems it's pointless for an application to block messenger backwards. And bruntly chaining the futures of `ms_dispatch_reset` events in turn cause the order between other real-time events incorrect. We might still need to understand if other types of events can block messenger/connection, but I don't think it is related to this PR.

For `ms_dispatch()`,  is its returned future always meaning to block reads? I feel the answer is no by looking at OSD code. For example `ms_dispatch()` is directly chaining the future returned by `OSD::handle_osd_map()`, meaning if we now decide to chain `ms_dispatch()` directly to the read path, the osdmap handling process would block all the other messages. Seems to me most of the futures returned by dispatching methods are only used to tracking the on-going continuations so that we can shutdown safely if they are gated.

From my understandings above, I think a possible way without unnecessary side-effects is continue to gate the returned futures by dispatching methods, and at the same time, do dispatching synchronously.

But we need to remember that there's a major effort left to unlink the futures that are not meant to block messenger/connections.

Reply:
For ms_dispatch_reset, I think the only important thing is that we get the reset before the new connection.  I can't think of a reason it would need to block the process, however.  Removing the future return value would seem to be fine.

It sounds to me like the osdmap handling path is wrong and needs to fork off an async process.

As I said above, I'm ok with simply restoring the future chaining within ChainedDispatcher.  I can audit the osd code once this is merged -- I don't actually think much needs to change.

Reply:
@xxhdx1985126 I'm not sure what you mean.

Reply:
@xxhdx1985126 Yes, as long as it doesn't block it will happen to work correctly.  My concern is that the interface suggests that adding a dispatcher that does block is safe, when it really isn't.  Thus, we might break this by accident in the future.  The consequence would be that an osd restarting at the wrong time would corrupt data some of the time randomly depending on timing.

For now, I think we can just modify ChainedDipatcher to return the future from do_for_each as before.  ProtocolV2 can simply drop them for now and I'll address it after this PR merges.

@xxhdx1985126 @cyx1231st Does this sound reasonable?

Reply:
@athanatos Sorry that I did't make myself clear. I mean I think the reason that we might not choose the way mentioned in your case 2 is that it may involve performance issues, so I think maybe we cam make some combination of your 2 cases. I think maybe we can follow your case 1 in implementing ms_dispatch while follow your case 2 in other ms_* methods. Am I right?

I'm ok with any final decision we make:)

Reply:
I don't think there's much chance for a performance issue either way.

Case 2 for is useful specifically for ms_dispatch because it allows downstream dispatchers to act as throttlers -- there are cases where this is attractive such as when a client has IO blocked on a specific object such that we want to prevent that specific tcp stream from making progress while permitting others.  Such a decision naturally rests down in the osd, not in the messenger, so it would be nice to let the dispatcher do it.

I'm with @cyx1231st that it's not obvious that an ms_handle_reset implementation (for instance) need *ever* block the invoker, so I'd suggest case 1 there.

Reply:
I think if ms_dispatch may block, there could a performance issue, right? As subsequent requests must wait for the current request to finish while it may be waiting for some external enevts, say a reply from another osd, right?

Reply:
Well, if we're using it for throttling, the *point* is to prevent another op from being read.

At the moment, the handler starts an async process *which it does not wait on*.  As such it does not actually block.  With that implementation, I think there would be at most a minimal difference from case 1.

The hope is that if the dispatcher does want to block, it can.  In the event that it doesn't there's no real performance downside to starting an async process and returning a seastar::now().  That way, we enable both usages.

Reply:
So I think what you suggest about ms_dispatch is that we keep it returning future<>, and if, in the future, someone what to make their implementation of ms_dispatch block, they should take care of the possible performance downgrade, right? Sorry for this back and forth discussion, just want to make sure that I follows:)

Reply:
There're 5 interfaces from `Dispatcher`. They need to be treated differently.

`ms_dispatch()` can go with case 2 b), and @athanatos will help to audit `ms_dispatch()` implementations in another PR to make sure its returned future really means to throttling.

Other interfaces are going with case 1 (return void), because currently messenger is expecting synchronous dispatching of them. (Although I think it is possible to make the future returned by `ms_handle_connect()` and `ms_handle_accept()` to delay the starting of message dispatching, but I don't think our current OSD code requires (or aware of) that, and it needs further subtle changes in `Protocol`).

I think this also means that `Dispatcher` shouldn't inherit `Gated`, OSD components should do that instead.

Reply:
@cyx1231st That all sounds right to me.  For now, 2a will suffice for ms_dispatch (as it was before), and I'll audit the users once this merges.

Reply:
> I think this also means that `Dispatcher` as part of messenger shouldn't inherit `Gated`, tracking non-messenger continuations are OSD components' responsibilities. (see [#34492 (comment)](https://github.com/ceph/ceph/pull/34492#discussion_r427706638))

@cyx1231st I think we better keep Dispatcher inherited from Gated, because this can serve as a note that any future Dispatchers should also consider guarding their ongoing continuations, am I right?

Reply:
@xxhdx1985126 I think the gate is meaningless for `Dispatcher`. For interfaces with case 1, there's nothing to gate as they are returning void as pure event notifications, and messenger doesn't care inside OSD components whether it means to spawn further async operations that need to be gated.

And for `ms_dispatch()`, it's returned future only means to throttling read, which should be strongly distinguished from the need of graceful shutdown.

Reply:
ok